Source/WebCore/accessibility/AXObjectCache.cpp

```diff
@@ -1715,9 +1715,26 @@ void AXObjectCache::onPageActivityStateChange(OptionSet<ActivityState> newState)
 #endif
 }
 
+static bool shouldDeferFocusNodeChange(Node* node)
+{
+    if (!node)
+        return false;
+
+    auto& document = node->document();
+    auto* renderer = node->renderer();
+
+    if ((renderer && renderer->needsLayout()) || document.needsStyleRecalc() || document.inRenderTreeUpdate() || (document.view() && document.view()->layoutContext().isInRenderTreeLayout()))
+        return true;
+
+    if (renderer && !renderer->beingDestroyed())
+        return true;
+
+    return false;
+}
+
 void AXObjectCache::onFocusChange(Node* oldNode, Node* newNode)
 {
-    if (nodeAndRendererAreValid(newNode) && rendererNeedsDeferredUpdate(*newNode->renderer())) {
+    if (shouldDeferFocusNodeChange(newNode)) {
```

Comment:
`shouldDeferFocusNodeChange` now seems very similar to `rendererNeedsDeferredUpdate`. Why does the former need to be added? Also, why remove the `nodeAndRendererAreValid` check?

Reply:
onFocusChange can be called before newNode has a renderer but document.needsStyleRecalc() and that is why we need to defer it.